From dd7c6661bd870f9ac52fce6f7f4f989248c12192 Mon Sep 17 00:00:00 2001
From: Paul Thomson <paulthomson@google.com>
Date: Mon, 27 Apr 2026 12:53:07 +0100
Subject: [PATCH 3/7] Exclude certain backend compiler modules for Lint

I believe this is just to reduce the size of the Kotlin jar (since these
backends should in theory never be needed). We also need to remove the
registration of extension points defined in the excluded modules.
---
 .../analysis-api/analysis-api-impl-base.xml   |  4 ----
 build.gradle.kts                              | 22 +++++++++++++++++++
 .../compiler/plugin/pluginRegisrar.kt         | 10 ---------
 .../src/main/kotlin/repoArtifacts.kt          | 12 ++++++++++
 4 files changed, 34 insertions(+), 14 deletions(-)

diff --git a/analysis/analysis-api-impl-base/resources/META-INF/analysis-api/analysis-api-impl-base.xml b/analysis/analysis-api-impl-base/resources/META-INF/analysis-api/analysis-api-impl-base.xml
index 8b77c3752fb..18c3a151a11 100644
--- a/analysis/analysis-api-impl-base/resources/META-INF/analysis-api/analysis-api-impl-base.xml
+++ b/analysis/analysis-api-impl-base/resources/META-INF/analysis-api/analysis-api-impl-base.xml
@@ -29,10 +29,6 @@
         interface="org.jetbrains.kotlin.analysis.api.symbols.KaAdditionalKDocResolutionProvider"
         dynamic="true"
     />
-
-    <extensionPoint qualifiedName="org.jetbrains.kotlin.irGenerationExtension"
-                    interface="org.jetbrains.kotlin.backend.common.extensions.IrGenerationExtension"
-                    area="IDEA_PROJECT" dynamic="true"/>
   </extensionPoints>
 
   <extensions defaultExtensionNs="com.intellij">
diff --git a/build.gradle.kts b/build.gradle.kts
index 82e84c8b73e..9d30f23aa4f 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -295,6 +295,28 @@ extra["compilerModules"] =
             commonCompilerModules +
             firAllCompilerModules
 
+// We exclude modules that include "backend".
+val lintExcludedModules = arrayOf(
+    ":compiler:ir.backend.common",
+    ":compiler:ir.backend.native",
+    ":compiler:backend.jvm",
+    ":compiler:backend.jvm.lower",
+    ":compiler:backend.jvm.codegen",
+    ":compiler:backend.jvm.entrypoint",
+    ":compiler:backend.js",
+    ":compiler:backend.wasm",
+    // Do not exclude. Required for `KotlinTypeMapper`.
+    // ":compiler:backend",
+    // Do not exclude. Required for `TypeMappingContext`.
+    // ":compiler:backend.common.jvm",
+
+    // These were previously excluded; we could perhaps exclude more,
+    // but it seems like backends are better candidates for exclusion.
+    ":compiler:ir.interpreter",
+    ":compiler:ir.serialization.js",
+    ":compiler:ir.serialization.jvm",
+).also { extra["lintExcludedModules"] = it }
+
 /**
  * An array of projects used in the IntelliJ Kotlin Plugin.
  *
diff --git a/plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/pluginRegisrar.kt b/plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/pluginRegisrar.kt
index e6482252643..3d7c6b52cc3 100644
--- a/plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/pluginRegisrar.kt
+++ b/plugins/scripting/scripting-compiler/src/org/jetbrains/kotlin/scripting/compiler/plugin/pluginRegisrar.kt
@@ -8,7 +8,6 @@
 package org.jetbrains.kotlin.scripting.compiler.plugin
 
 import com.intellij.mock.MockProject
-import org.jetbrains.kotlin.backend.common.extensions.IrGenerationExtension
 import org.jetbrains.kotlin.cli.common.extensions.ReplFactoryExtension
 import org.jetbrains.kotlin.cli.common.extensions.ScriptEvaluationExtension
 import org.jetbrains.kotlin.cli.common.extensions.ShellExtension
@@ -81,15 +80,6 @@ class ScriptingCompilerConfigurationComponentRegistrar : ComponentRegistrar {
             SyntheticResolveExtension.registerExtension(project, ScriptingResolveExtension())
             ExtraImportsProviderExtension.registerExtension(project, ScriptExtraImportsProviderExtension())
 
-            with(configuration.extensionsStorage!!) {
-                if (configuration.get(ENABLE_SCRIPT_EXPLANATION_OPTION, false)) {
-                    IrGenerationExtension.registerExtension(ScriptingIrExplainGenerationExtension(project))
-                }
-
-                IrGenerationExtension.registerExtension(ScriptLoweringExtension())
-                IrGenerationExtension.registerExtension(ReplLoweringExtension())
-            }
-
             if (messageCollector != null) {
                 project.registerService(ScriptReportSink::class.java, CliScriptReportSink(messageCollector))
             }
diff --git a/repo/gradle-build-conventions/buildsrc-compat/src/main/kotlin/repoArtifacts.kt b/repo/gradle-build-conventions/buildsrc-compat/src/main/kotlin/repoArtifacts.kt
index 7e65bfcf69f..9a2acced4b8 100644
--- a/repo/gradle-build-conventions/buildsrc-compat/src/main/kotlin/repoArtifacts.kt
+++ b/repo/gradle-build-conventions/buildsrc-compat/src/main/kotlin/repoArtifacts.kt
@@ -304,8 +304,20 @@ fun Project.publishJarsForIde(
     libraryDependencies: List<String> = emptyList(),
     jarTaskConfiguration: Jar.() -> Unit = {},
 ) {
+    publishJarsForIdeHelper(projects, libraryDependencies, jarTaskConfiguration)
+}
+
+private fun Project.publishJarsForIdeHelper(
+    projectsFull: List<String>,
+    libraryDependencies: List<String> = emptyList(),
+    jarTaskConfiguration: Jar.() -> Unit = {},
+) {
+
     val projectsUsedInIntelliJKotlinPlugin: Array<String> by rootProject.extra
 
+    val lintExcludedModules: Array<String> by rootProject.extra
+    val projects = projectsFull - lintExcludedModules
+
     for (projectName in projects) {
         check(projectName in projectsUsedInIntelliJKotlinPlugin) {
             "`$projectName` is used in IntelliJ Kotlin Plugin, it should be added to `extra[\"projectsUsedInIntelliJKotlinPlugin\"]`"
-- 
2.54.0.545.g6539524ca2-goog

